Remove packages folder from Compile and EmbeddedResource item lists

dsplaisted opened the pull request:
Fixes #306

dsplaisted committed:
Remove packages folders from Compile and EmbeddedResource item lists, but allow opt-out

Fixes #306

dsplaisted commented:
Should we have any sort of testing for the default excludes?  I didn't find any.

Also, repeating my [comment on the issue](https://github.com/dotnet/sdk/issues/306#issuecomment-255491061): What about `node_modules` or similar folders?  I thought one of the requirements for decent globbing perf was to avoid looking at those folders, but I don't see them excluded anywhere.

nguerrera commented:
Tests would definitely be good. ;)

nguerrera commented:
node_modules makes sense. Any other package manager conventions we should add?

livarcocc commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets:
```diff
@@ -18,7 +18,7 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   <PropertyGroup>
-    <DefaultExcludes>bin\**;obj\**</DefaultExcludes>
+    <DefaultExcludes>bin\**;obj\**;packages\**</DefaultExcludes>
```
could we add **\*.xproj as well? This would match exactly what we currently have in project.json world.

dsplaisted replied:
@livarcocc The only case where adding `**\*.xproj` here would have an effect is if there is an `.xproj` in under the `compiler\resources` folder.  Is there any normal scenario where that would happen?

livarcocc replied:
I see, so this does not get pulled into the excludes for compile and embedded resources? That's where we have this exclude in the CLI.

nguerrera replied:
It does apply to those but the corresponding includes will be filtered by extension (*.cs, *.vb, *.resx). Those won't match xproj. The one exception is that project.json included compiler/resources/**/* (no extension filter) as resources. That's why it's the only case where the xproj exclude matters.

nguerrera commented:
Also xprojs will cease to exist, right? We aren't doing future maintainers a favor if they end up staring at a default about format that never hit 1.0. If you want 100% guaranteed compat on that, then add the xproj exclude to the compiler/resources include in migration. I'd rather this relic live in migrated projects alone than in defaults for the rest of time.

livarcocc commented:
We have it in the migration, but I would like to remove it from there so that those projects are clean(er).

I was just working under the assumption from this issue here opened by Sri: https://github.com/dotnet/cli/issues/4330. Item 5 is about excludes.

nguerrera commented:
I think having xproj mentioned in these targets is instant legacy. Won't folks want to delete them anyway? Maybe just leave them included on migration? Can we get data about how often this exclude actually triggers in practice?

srivatsn commented:
I agree we shouldn't mention xprojs in the exclude. Migration would convert and delete those files where applicable and so we shouldn't run into any cases that need it.

livarcocc commented:
Ok. We will move the xprojs over to a backup folder and I WILL NOT add this exclude to the excludes during migration.

Given that these are people moving from project.json to csproj, I want to keep the csproj to a bare minimum of verbosity.

dsplaisted commented:
I've added `node_modules` and `jspm_packages` per feedback from the web team, and added a way to opt out of removing these files if you do want to include something from any of these folders.

eerhardt commented:
> I've added node_modules and jspm_packages per feedback from the web team

Why can't the web sdk add removes for these folders?  Why should the core .NET SDK care about npm or jspm?  Those are web project concerns.  The can put them in the web sdk nupkg, like they did in the folloiwng:  https://github.com/dotnet/sdk/pull/223/files

srivatsn commented:
Agree with Eric. Also don't they already have excludes for content, they could add for Compile as well.

dsplaisted commented:
OK, I've sent a PR to the web Sdk to remove the node and jspm package folders from the item lists.  What about the `packages` folder.  Should we remove that in the .NET Sdk, or in the Web Sdk?

eerhardt commented:
> What about the packages folder. Should we remove that in the .NET Sdk, or in the Web Sdk?

Since `packages` is typically NuGet's folder, I'm fine with the .NET Sdk excluding it.  We already know/care about NuGet in the .NET Sdk.

dsplaisted committed:
Add GetValuesCommand to save property or item values to a file

dsplaisted committed:
Add tests for excluded folders

dsplaisted commented:
I've updated this based on feedback and added tests.  It should be ready for review now.

dsplaisted committed:
Fix excluded folder tests on non-Windows

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs:
```diff
@@ -66,6 +68,109 @@ public void It_builds_the_library_twice_in_a_row()
+
+            compileItems = compileItems.Where(i =>
+                    !i.EndsWith("AssemblyAttributes.cs", System.StringComparison.OrdinalIgnoreCase) &&
+                    !i.EndsWith("AssemblyInfo.cs"))
```
Why do these need to be special cased? Also, why are there two files?

dsplaisted replied:
One is for the `TargetFrameworkAttribute` and comes from the temp folder.  The other is for the other assembly attributes that the SDK adds, and is in the intermediate output path.

I've factored this out into a separate method that removes these, with a comment to explain what's going on.

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs:
```diff
@@ -66,6 +68,109 @@ public void It_builds_the_library_twice_in_a_row()
+            compileItems = compileItems.Where(i =>
+                    !i.EndsWith("AssemblyAttributes.cs", System.StringComparison.OrdinalIgnoreCase) &&
+                    !i.EndsWith("AssemblyInfo.cs"))
+                .ToList();
```
Same question, can you comment and factor out a helper?

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs:
```diff
@@ -66,6 +68,109 @@ public void It_builds_the_library_twice_in_a_row()
+            .Select(item => item.Replace('\\', Path.DirectorySeparatorChar))
+            .ToArray();
+
+
```
nit: extra line of whitespace

dsplaisted committed:
Factor out RemoveGeneratedCompileItems method and and comment explaining what's going on

dsplaisted commented:
@srivatsn @MattGertz for approval

#### Scenario 
Parity in behavior with project.json projects, which automatically ignored files in the packages folder.

#### Bug
#306

#### Workarounds
You could add a `Remove` or `Exclude` yourself in the project file.  But this would mean that `dotnet migrate` would have to do so which would make the migrated projects uglier than they need to be.

#### Risk
Low risk - this only applies to projects using the .NET SDK.

#### Performance Impact
If anything it should improve performance as if there's a packages folder it won't need to be scanned for files that match a glob.

#### Regression Analysis
n/a

MattGertz commented:
@dsplaisted Looks good, once Sri signs off I will approve.

MattGertz commented:
Approved.
